Import tests from Google V8 (generator syntax)

jugglinmike opened the pull request:
These tests are derived from the following files within the Google V8
project:

    test/mjsunit/es6/generators-parsing.js

I have deliberately omitted one assertion from the original test file because I believe it to be invalid:

```js
assertThrows("function* yield() { \"use strict\"; (yield 3) + (yield 4); }",
             SyntaxError);
```

(source: https://github.com/v8/v8-git-mirror/blob/3a79259c0d4ae2d314bb57ad2d8dcde8f2478974/test/mjsunit/es6/generators-parsing.js#L96-L97)

[I've reported this to the V8 team by filing a bug against that project](https://code.google.com/p/v8/issues/detail?id=3975)

arv commented on test/language/generators/syntax/return.js:
```diff
@@ -0,0 +1,18 @@
+
+function* g() { return; }
+function* g() { return 1; }
+function* g() {
```
Not clear why this is repeated?

jugglinmike replied:
I suppose to account for an edge case where a newline would trigger different parsing behavior. That seems more like a regression test, though. If we were to try to catch all problems like this exhaustively, we'd be here all day! I'll remove all the duplicates you've identified.

arv commented on test/language/generators/syntax/yield-as-binding-identifier.js:
```diff
@@ -0,0 +1,14 @@
+/*---
+  description: >
+      `yield` is a reserved keyword within generator function bodies and may
+      not be used as a binding identifier.
```
This description is not correct. In this case `yield` is an `IdentitfierExpression`.

jugglinmike replied:
I'm not seeing `IdentifierExpression` in the spec. Was this maybe a typo?

arv commented on test/language/generators/syntax/yield-as-computed-property-name.js:
```diff
@@ -0,0 +1,13 @@
+
+/*---
+  description: >
+      `yield` may be used as a computed property name in an object literal
```
s/computed property name/literal property name/

jugglinmike replied:
Thanks--renamed the file to match.

arv commented on test/language/generators/syntax/yield-as-expression-with-rhs.js:
```diff
@@ -0,0 +1,19 @@
+function* g() { {yield 1} }
+function* g() { yield 1, yield 1; }
+function* g() { (yield 1) ? yield 1 : yield 1 }
+function* g() {
```
this is another repeat

arv commented on test/language/generators/syntax/yield-as-statement.js:
```diff
@@ -0,0 +1,18 @@
+
+function* g() { yield; }
+function* g() { yield 1; }
+function* g() {
```
repeat?

arv commented on test/language/generators/syntax/yield-as-yield-operand.js:
```diff
@@ -0,0 +1,13 @@
+
+/*---
+  description: >
+      `yield` expressions may be used as the right-hand-side of other `yeild`
```
typo

s/yeild/yield/

jugglinmike replied:
Thanks

anba commented on test/language/generators/syntax/yield-as-function-expression-binding-identifier.js:
```diff
@@ -0,0 +1,14 @@
+ ---*/
+
+function* g() {
+  (function yield() {})
```
This is actually allowed: 
> FunctionExpression :
> &nbsp;&nbsp;function BindingIdentifier<sub>opt</sub> ( FormalParameters ) { FunctionBody }

BindingIdentifier has not [Yield] production parameter, so `yield` is a valid identifier in this context.

jugglinmike replied:
And here I thought I was getting good at reading the spec,,, Nice catch!

I've written up a bug report for V8. Since you identified the problem, I thought I'd check with you first: would you like to report it instead?

anba replied:
> would you like to report it instead?

No, I don't mind if you report this.

jugglinmike replied:
Got it! You can follow along here if you like: [Issue 3983: Incorrect restriction on binding identifiers of function expressions](https://code.google.com/p/v8/issues/detail?id=3983)

bterlson commented:
Please remove the syntax subdirectory many of these are under.

jugglinmike commented:
The "syntax" subdirectory may seem extraneous in the context of this pull request, but I have organized the tests like this to better accommodate the semantics-related tests defined in these files:

- [v8/test/mjsunit/es6/generators-iteration.js](https://github.com/v8/v8-git-mirror/blob/8e4e83119ca187e8c0e2a3b518cac2dc3def3fa3/test/mjsunit/es6/generators-iteration.js)
- [v8/test/mjsunit/es6/generators-objects.js](https://github.com/v8/v8-git-mirror/blob/8e4e83119ca187e8c0e2a3b518cac2dc3def3fa3/test/mjsunit/es6/generators-objects.js)
- [v8/test/mjsunit/es6/generators-runtime.js](https://github.com/v8/v8-git-mirror/blob/8e4e83119ca187e8c0e2a3b518cac2dc3def3fa3/test/mjsunit/es6/generators-runtime.js)
- [v8/test/mjsunit/es6/generators-states.js](https://github.com/v8/v8-git-mirror/blob/8e4e83119ca187e8c0e2a3b518cac2dc3def3fa3/test/mjsunit/es6/generators-states.js)

Of course, I'm happy to use whatever organization you like; I just wanted to make sure my intentions were clear before updating the patch.

bterlson commented:
Hmm, thanks for the clarification. I see what you're doing... here's my thoughts, but I have just now begun thinking about this so it could be dumb :)

generators-iteration and generators-states are exploring various usages of yield. I think this belongs in language/generators (or potentially language/yield, although I don't really think that distinction is necessary or desirable).

Many tests in generators-objects and generators-runtime are testing shapes of built-ins so I would place these under built-ins/Generator and built-ins/Generator/prototype as appropriate.

One oddity is that I guess is that tests for the shape of the object returned from calling a generator should be under language/generators.

Thoughts?

The reason I want to make the organization clean is that the organization is also basically a test tagging scheme so I want to be very clear and deliberate with the organizational scheme. Eventually it will be reflected on our new website so the simpler it is the easier the mapping will be in the future.

bterlson commented:
Test content looks good. The only other comment I have is to suggest that language/generators be renamed to language/statements/generator, and have every test duplicated for a generator expression (put those under language/expressions/generator).

jugglinmike commented:
@bterlson sounds good to me! I've re-named the tests as you've requested.

I've added the GeneratorExpression analogs along with copies for GeneratorMethods declared within object literals and within class bodies. I plan on preserving those additions as separate commits in the final version of this pull request (since they did not come from V8 directly).

One thing that occurs to me now is that most of the "positive syntax" tests don't actually assert anything. For an example of one file that *does* include an assertion, see [the proposed `/language/expressions/generators/no-yield.js`](https://github.com/bocoup/test262/blob/6e5fdbebff436e41a49999090aaf2015ac933879/test/language/expressions/generators/no-yield.js). I'd like to add a basic assertion along those lines to each of the positive syntax tests so they provide some minimal assurance that the runtime actually executes the code that it parsed. Does that sound good to you?

bterlson commented:
> One thing that occurs to me now is that most of the "positive syntax" tests don't actually assert anything. 

I thought about this too. I could go either way but I think they're fine as-is. test262-harness verifies that the test runs to completion by appending `;$DONE();` to the end of the test file. I suppose you could do more in-depth validation to ensure the syntax results in a value you expect, but this seems of questionable utility to me. So, IOW, I think the tests are fine as is, but if you want to do more in depth validations I certainly won't object :)

jugglinmike commented:
`yield` is not a valid identifier in generator function bodies, regardless of whether the code is strict mode or not. This means that the following tests are invalid in their current form:

- test/language/expressions/generators/yield-as-generator-expression-binding-identifier.js
- test/language/method-definition/class/generators/yield-as-generator-method-binding-identifier.js 
- test/language/method-definition/object/generators/yield-as-generator-method-binding-identifier.js 
- test/language/statements/generators/yield-as-generator-declaration-binding-identifier.js

I will remove the `strictOnly` flag and add `negative: SyntaxError` in all cases

jugglinmike commented:
After a bit more time studying the subscripting notation (and with thanks to @caitp and @arv), I've realized that only the first test I referenced above (yield-as-generator-expression-binding-identifier.js) is invalid.

From [14.4 Generator Function Definitions](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-generator-function-definitions-static-semantics-early-errors):

> *GeneratorExpression* :
>    **function** * *BindingIdentifier*<sub>[Yield]opt</sub> *FormalParameters*<sub>[Yield,GeneratorParameter]</sub> ) { *GeneratorBody* }

...and from [12.1 Identifiers](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-identifiers):

> *BindingIdentifier*<sub>[Yield]</sub> :
>     Identifier
>     [~Yield] **yield**

...and [5.1.5 Grammar Notation](https://people.mozilla.org/~jorendorff/es6-draft.html#sec-grammar-notation):

> If a right-hand side alternative is prefixed with “[+parameter]” that
> alternative is only available if the named parameter was used in
> referencing the production’s nonterminal symbol. If a right-hand side
> alternative is prefixed with “[~parameter]” that alternative is only
> available if the named parameter was not used in referencing the
> production’s nonterminal symbol.

The `[Yield]` subscript modifies the definition of `BindingIdentifier` by restricting it to the set defined by `Identifier`, so `yield` is never a valid *GeneratorExpression* name.

The three other tests I referenced concern *GeneratorMethod*s and *GeneratorDeclaration*s. Section 14.4 defines each of these with "*BindingIdentifier*<sub>[?Yield]</sub>", meaning `yield` *is* a valid binding identifier in those cases.

jugglinmike commented:
Okay @bterlson, I've pushed up a few more commits to fix the bugs and add basic assertions.

Prior to landing, I intend to squash these into three commits:

1. Import tests from Google V8 (generator syntax)
  - fixup! Remove repeated assertions
  - fixup! Correct typos
  - fixup! Remove repeated assertion
  - fixup! Correct faulty test
  - fixup! Rename tests
 - fixup! Limit Generator statement to non-strict
2. Add asertions for Generator Statement syntax tests
3. Add tests for additional generator function forms
  - Add tests for generator expressions
  - Add tests for generator method definitions
  - fixup! Limit tests to non-strict
  - fixup! Correct invalid test
  - fixup! Add basic tests for positive syntax

Let me know what you think!

jugglinmike commented:
As per our discussion in IRC, I've made the following changes:

- Moved `language/identifiers/syntax/*` -> `language/identifiers/`
- Moved `language/method-definition/class/generators/*` -> `language/statements/class/method-definition/*` (renaming `return.js` to `generator-return.js` and `no-yield.js` to `generator-no-yield.js`)
- Moved `language/method-definition/object/generators/*` -> `language/expressions/object/method-definition/*` (renaming `return.js` to `generator-return.js` and `no-yield.js` to `generator-no-yield.js`)
- Added  `features: [generators]` to the frontmatter of the "method-definition" tests

bterlson commented:
Great work landing this!!

jugglinmike commented:
Thanks for your help :)
